### DISCUSSION TITLE
Trying to build for aarch64

### DISCUSSION THREAD

**Discussion by screech** [eythaann/Seelen-UI#209] [Q&A]
Cloned the directory, followed instructions and ran
"npm install && npm run dev"

got this error on npm install

:Seelen-UI/ (master) $ npm install
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path <home>/Seelen-UI/node_modules/readable-types/rtt-plugin/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '<home>/Seelen-UI/node_modules/readable-types/rtt-plugin/package.json'

Total newbie to this kind of code, more familiar with systems software/kernel etc .. any pointers would be appreciated

**Comment by eythaann** [comment]
readable types is just type library than I made some months ago before start Seelen UI project, so idk why it fails on npm install in your case.
It looks like npm is not handling routed libraries on your system.

**Reply by screech** [reply]
Is there a workaround I can use like editing a package.json and preventing NPM from looking for it, or is this critical and I need to get NPM to do the right thing ?

FWIW I tried cloning the rtt-plugin repo in the right place in the dir structure and I got a "corrupted tarball" on npm install

**Comment by screech** [comment]
So figured it out. WSL was getting in the way.

Installing packages and doing it in Windows Native (CMD) shell fixed most issues.

1. Had to comment out the SSE instructions (will port them to ARM neon next)
2. npm run dev is starts (see below) and then after a while craps out saying it tried to allocate 1TB of memory and failed. Any suggestions @eythaann ?

warning: `seelen-ui` (bin "seelen-ui") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 15.26s
[INFO~][seelen_ui] ───────────────────── Starting Seelen UI v2.0.4 ─────────────────────
[INFO~][seelen_ui] Operating System: Windows 10.0.26120 (Windows 11 Core) [64-bit]
[INFO~][seelen_ui] WebView2 Runtime: Ok("130.0.2849.68")
[INFO~][seelen_ui] Elevated        : Ok(false)
[INFO~][seelen_ui] Locate          : Some("en-US")
[INFO~][seelen_ui::modules::cli] TCP server listening on 127.0.0.1:50743
[TRACE][seelen_ui::seelen] Initializing Seelen
[TRACE][seelen_ui::seelen] Ensuring folders
[TRACE][seelen_ui::state::application] Creating new State Manager

**Reply by eythaann** [reply]
I don't have any idea, a user already work on a port to arm64 but It was never finished I think, I haven't worked with arm processor before so also I don't have experience migrating apps to arm, btw this will be a good first time.

**Reply by HgGamer** [reply]
Hi, 
I'm working on arm64 port, the 2.0 beta branch was working on arm with a little fix.

**Reply by screech** [reply]
I don't see a beta branch in github, do you have a URL I can pull from ?

**Reply by HgGamer** [reply]
it was merged back to master, as soon as I have the free time I will finish the port, if you want to do it yourself fix the icon_extractor.rs
platform dependency and go back in master a month or two you will be able to build the project, then you can merge back the commits, Im currently facing an issue where the icons are not collected properly and it crashes.

**Reply by screech** [reply]
the SSE code in icon_extractor I have currently commented out and like I said I can move to neon from sse unless u have that already.

Ok will try a date from a month ago (if there is a commit I can look at that would be awesome) but the built part is not starting for me ..

**Reply by eythaann** [reply]
@screech do you know a way to work on a remote ARM computer? I don't wanna buy a new laptop only for a couple weeks of development haha.

**Reply by screech** [reply]
nope .. running on my local desktop. There are Mac's that you can get on
the cloud, but dont know of any windows platforms :(

**Comment by screech** [comment]
I hacked enough to get to here
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 29.78s
[INFO~][seelen_ui] ───────────────────── Starting Seelen UI v2.0.4 ─────────────────────
[INFO~][seelen_ui] Operating System: Windows 10.0.26120 (Windows 11 Core) [64-bit]
[INFO~][seelen_ui] WebView2 Runtime: Ok("130.0.2849.68")
[INFO~][seelen_ui] Elevated        : Ok(false)
[INFO~][seelen_ui] Locate          : Some("en-US")
[INFO~][seelen_ui::modules::cli] TCP server listening on 127.0.0.1:60698
[TRACE][seelen_ui::seelen] Initializing Seelen
[TRACE][seelen_ui::seelen] Ensuring folders
[TRACE][seelen_ui::state::application] Creating new State Manager
[TRACE][seelen_ui::state::application] Starting Seelen UI Files Watcher
[INFO~][seelen_ui::seelen_wall] Creating seelen-wall
[TRACE][seelen_ui::seelen] Enumerating Monitors & Creating Instances
[INFO~][seelen_ui::seelen_bar] Creating fancy-toolbar/DISPLAY1
[INFO~][seelen_ui::seelen_weg] Creating seelenweg/DISPLAY1
[DEBUG][seelen_ui::utils] lazy setup start
[TRACE][seelen_ui::hook] Registering Windows and Virtual Desktop Hooks
[ERROR][src/background/main.rs:69] A panic occurred:
  Cause: SEELEN mutex is deadlocked
  Location: src/background\hook.rs:221:26
[ERROR][src/background/main.rs:69] A panic occurred:
  Cause: panic in a function that cannot unwind
  Location: library\core\src\panicking.rs:221:5
thread caused non-unwinding panic. aborting.
[nodemon] app crashed - waiting for file changes before starting...

Anything familiar here ?

I commented load_icon_pack out as that was the main problem in aarch64

**Reply by screech** [reply]
Failing to get past this mutex .. 
```
    ($mutex:expr, $duration:expr) => {{
        let guard = $mutex.try_lock_for(std::time::Duration::from_secs($duration));
        let guard_name = stringify!($mutex);
        if guard.is_none() {
            let mut panic_msg = format!("{} mutex is deadlocked", guard_name);
```

And no guard exists, But not sure why we ended up here in the 1st place. Any thoughts would be appreciated

**Reply by eythaann** [reply]
It is a macro, the error will tell you where is called in this case the win hook is been blocked, by some other blocker action that uses the SEELEN singleton struct on the get icons function.

**Reply by screech** [reply]
so finally had time to chase this down.

The issue was that the Seelen mutex was not initialized as other init threads were not hanging/stalling.

Hacked my way through and got it to complete startup from an optimized ARM64 build (which feels snappier). The things I had to comment out as they were the calls hanging are in the commit here.

https://github.com/screech/Seelen-UI/commit/e2d2b4ac2b25ffa97346500961e302bfd4e57f6e

If I have a little more time, will chase why they are hanging for me .. maybe too many desktops/windows ?

**Reply by eythaann** [reply]
@screech this sometimes happen on x64 so is not an ARM issue I think, maybe?

**Reply by screech** [reply]
This is one of the places where I cannot get anything to work
   >> get_vd_manager().listen_events(sender)?;

I can't seem to figure out why this call never returns

**Comment by screech** [comment]
@eythaann some more info. Synced with the latest 2.0.12 and now I see only two issues (the rest seems to be all good)
#1 : In seelen_weg/mod.rs 
`let window = Window::from(hwnd);` is not returning and just hangs the thread
#2 : in hook.rs 
`get_vd_manager().listen_events(sender)?;` also hangs the thread

Both seem to be calling in the Windows:win32 crate .. not at all familiar with this space, could be that ARM64 needs a different crate ? Even a hint for what to search for would help

FWIW : installing the x64 2.0.12 on my ARM box now crashes, whereas up 2.0.6 it worked fine in emulation mode (sometimes a little slow)

**Comment by Magniveo** [comment]
How build its on windows 10 x64? 
After `npm run dev` error 
Caused by:
  the cargo feature `profile-rustflags` requires a nightly version of Cargo, but this is the `stable` channel
  See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
  See https://doc.rust-lang.org/cargo/reference/unstable.html#profile-rustflags-option for more information about using this feature."
